WithNames format fixes

azat opened the pull request:
Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix missing or excessive headers in `TSV/CSVWithNames` formats. This fixes #12504


- Fix `File(TSVWithNames*)` (header was written multiple times)
- Fix `clickhouse-local --format CSVWithNames*` (lack of header, broken after #12197)
- Fix `clickhouse-local --format CSVWithNames*` with zero rows (lack of header)

Continuation of: #12197
Fixes: #12504

<details>

HEADs:
- 8db28001e3d903bf824788ae989d888dd8db5bcb
- f58e4bf96a56aeffa2067d114b506fa9b3ef7c72

</details>

azat committed:
Add csv_with_names prefix to 01375_storage_file_write_prefix test

azat committed:
Add test for File(TSVWithNames)

azat committed:
Write header only once in TSVWithNames*

azat committed:
Call doWritePrefix() if IOutputFormat works as Processor (i.e. clickhouse-local)

This fixes the *WithNames output format (lack of header before this
patch).

azat committed:
Add a test for *WithNames* with clickhouse-local

akuzm commented on tests/queries/0_stateless/01375_output_format_with_names.sh:
```diff
@@ -0,0 +1,22 @@
+}
+
+echo 'TSVWithNames'
+produce | ${CLICKHOUSE_LOCAL} "${opts[@]}" --format TSVWithNames
```
Is the `produce` part important, or `select '1' key from numbers(2) format TSVWithNames ` would also work?

azat replied:
egh, was too enthusiastic about scripting, will use your suggestion

akuzm commented:
@nikitamikhaylov FYI
You'll probably have to change all this when adding parallel formatting, so it's good that we have more test cases now.

nikitamikhaylov commented:
> @nikitamikhaylov FYI
> 
> You'll probably have to change all this when adding parallel formatting, so it's good that we have more test cases now.

Thanks! I will think.

akuzm commented:
I'll take a look at the perf test -- probably it depends on the old broken behavior and doesn't expect the header somewhere.

akuzm commented:
The reason the performance test fails is that now we can't read an empty file as TSVWithNamesAndTypes:

```
$ clickhouse-local --query "
   create table t engine File(TSVWithNamesAndTypes, '1.tsv') as
      select * from numbers(1) where number < 0;
   select * from t;
"

Code: 117, e.displayText() = DB::Exception: Unknown field found in TSV header: '' at position 0
Set the 'input_format_skip_unknown_fields' parameter explicitly to ignore and proceed: Error while reading from '1.tsv': While executing File (version 20.7.1.1)
```

Can we add some support for this?

azat commented:
>The reason the performance test fails is that now we can't read an empty file as TSVWithNamesAndTypes:

Not really, the problem is that the header wasn't written, while it should.

P.S. and reading empty files is not supported anyway, this is the result w/o this PR:

```sql
SELECT * FROM file('/dev/null', TSVWithNamesAndTypes, 'key Int')
Received exception from server (version 20.6.1):
Code: 117. DB::Exception: Received from localhost:9000. DB::Exception: Unknown field found in TSV header: '' at position 0
Set the 'input_format_skip_unknown_fields' parameter explicitly to ignore and proceed: While executing File.
```

azat committed:
Call writePrefix() from ctor in SinkToOutputStream

This will fix lack of headers for the following formats if there are
zero rows:
- TSVWithNames
- CSVWithNames
- TSVWithNamesAndTypes

azat committed:
Add a test for *WithNames* via clickhouse-local (corner cases)

- empty rows
- multiple runs to the same file (write header only once)

azat commented:
>the problem is that the header wasn't written, while it should.

@akuzm fixed

artbeglaryan commented:
@akuzm will this be backported to 20.3 too? On original task https://github.com/ClickHouse/ClickHouse/pull/12197 it marked as 20.3-backported

```
artur@sv-marmoset206:~$ clickhouse-client
ClickHouse client version 20.3.17.173 (official build).
Connecting to localhost:9000 as user default.
Connected to ClickHouse server version 20.3.17 revision 54433.

sv-marmoset206.maas :) Bye.
artur@sv-marmoset206:~$ curl  http://localhost:8123/?query=SELECT+1+as+T+FORMAT+CSVWithNames
1
```

Also didn't work on latest 20.4 too

20.4
```
artur@sv-marmoset206:~$ clickhouse-client
ClickHouse client version 20.4.9.110 (official build).
Connecting to localhost:9000 as user default.
Connected to ClickHouse server version 20.4.9 revision 54434.

sv-marmoset206.maas :) Bye.
artur@sv-marmoset206:~$ curl  http://localhost:8123/?query=SELECT+1+as+T+FORMAT+CSVWithNames
1
```

20.5 is ok
```
artur@sv-marmoset206:~$ clickhouse-client
ClickHouse client version 20.5.5.74 (official build).
Connecting to localhost:9000 as user default.
Connected to ClickHouse server version 20.5.5 revision 54435.

sv-marmoset206.maas :) Bye.
artur@sv-marmoset206:~$ curl  http://localhost:8123/?query=SELECT+1+as+T+FORMAT+CSVWithNames
"T"
1
```
